fix(Sharding): strict type context and return

kyranet opened the pull request:
**Please describe the changes this PR makes and why it should be merged:**

Right now, `ShardClientUtil#broadcastEval` and `ShardingManager#broadcastEval`'s return types are the same as the callback's, as well as the context passed, however, the context data is serialized, and so is the return type.

This PR implements `Serialized`, which fixes this issue by converting rich types into their serialized counterpart.

**Status and versioning classification:**

- I know how to update typings and have done so, or typings don't need updating

kyranet commented:
I'm unsure how to properly type the returns because of the worker mode:

https://github.com/discordjs/discord.js/blob/985d4d6a438fc9123264f6a1c600f34fccc1825f/src/sharding/ShardClientUtil.js#L92-L95

[`process.send`](https://nodejs.org/api/process.html#process_process_send_message_sendhandle_options_callback) seems to JSON serialize values, however, the worker mode supports sending a lot richer data, as this is powered by V8's serder instead.

iCrawl commented:
This needs a rebase.

iCrawl commented:
Can't we just go with this for now, despite the worker mode and potentially wait for an issue in case anyone is using this to send more complicated data?

LosTigeros commented:
What about executing functions that would return a Promise on the shard?
Can't [this](https://github.com/discordjs/discord.js/blob/master/src/sharding/ShardingManager.js#L285) be like:
```js
promises.push(Promise.resolve(sh[method](...args)));
```
This would work for normal functions and PromiseLike ones so `broadcastEval` `script` argument could be either `T` or `Promise<T>`

kyranet commented:
Yes, we can go for this for now, and if we wanted to make both consistent, we can always use [V8's serialization API](https://nodejs.org/dist/latest-v16.x/docs/api/v8.html#v8_serialization_api), which is what worker threads use, @iCrawl.

Nice catch @LosTigeros! I changed the return of the callbacks from `T` to `Awaited<T>`, so in case it's a promise, the promise's type is unwrapped. Also added tests to make sure they work as expected, see 52ec31f :)
